test/suites/image.sh

```diff
@@ -398,3 +398,93 @@ run_images_public() {
   lxc project delete foo
   lxc image delete "${fingerprint}"
 }
+
+test_image_cached() {
+  echo "==> Test that images are being cached consistently between projects."
+
+  if lxc image alias list testimage | grep -wF "testimage"; then
+    lxc image delete testimage
+  fi
+
+  echo "==> Set up a local LXD server as an image remote."
+  local LXD2_DIR LXD2_ADDR
+  LXD2_DIR=$(mktemp -d -p "${TEST_DIR}" XXX)
+  spawn_lxd "${LXD2_DIR}" true
+  LXD2_ADDR=$(< "${LXD2_DIR}/lxd.addr")
+
+  echo "==> Use busybox image for testing."
+  LXD_DIR="${LXD2_DIR}" deps/import-busybox --alias testimage --public
+
+  token="$(LXD_DIR=${LXD2_DIR} lxc config trust add --name foo -q)"
+  lxc remote add l2 "${LXD2_ADDR}" --token "${token}"
+
+  echo "==> Create a new instance in the default project with a remote image that is not stored locally."
+  lxc init l2:testimage c1
+
+  local fingerprint
+  fingerprint="$(lxc config get c1 volatile.base_image)"
+
+  echo "==> Delete the instance."
+  lxc delete c1
+
+  echo "==> Check that the locally saved image is marked as cached."
+  lxc image info "${fingerprint}" | grep -xF "Cached: yes"
+
+  echo "==> Explicitly copy the same remote image to local storage."
+  lxc image copy l2:testimage local:
+
+  echo "==> Check that the locally saved image is now marked as non-cached."
+  lxc image info "${fingerprint}" | grep -xF "Cached: no"
+
+  echo "==> Delete the local image."
+  lxc image delete "${fingerprint}"
+
+  echo "==> Create a new project p1. It has features.images=true by default."
+  defaultPoolName="$(lxc profile device get default root pool)"
+  lxc project create --storage "${defaultPoolName}" p1
+
+  echo "==> Switch to project p1."
+  lxc project switch p1
```

Comment:
For tests I recommend using `--project` flag rather than persistent switching as its clearer when reviewing the tests which project each action is occurring in rather than having to remember the state.